Wavelet transform analysis of 1-D signals using python (#6757)
section-engineering-education/engineering-education: content/articles/wavelet-transform-analysis-of-1d-signals-using-python/index.md

```python
sigma = 0.05  # Noise variance
x_noisy = x + sigma * np.random.randn(x.size)   # Adding noise to signal

# Wavelet denoising
x_denoise = denoise_wavelet(x_noisy, method='VisuShrink', mode='soft', wavelet_levels=3, wavelet='sym8', rescale_sigma='True')
```

After adding the random noise, we denoise the signal using the `denoise_wavelet()` function. Here, we use the `VishuShrink` method. 
For visualization, we plot the output.

```python
plt.figure(figsize=(20, 10), dpi=100)
plt.plot(x_noisy)
plt.plot(x_denoise)
plt.show()
```

![Output for the denoised audio signal](/engineering-education/wavelet-transform-analysis-of-1d-signals-using-python/wavelet-three.png)

The blue signal is the noisy signal, and the orange is the denoised output. Now, it is hard to know whether the denoising is done or not. It can be done by either calculating the peak signal to noise ratio(PSNR) or physically listening to the music. 

We add the `sounddevice` python package to anaconda to listen to the audios. All these can be done in the terminal. For example, to import `sounddevice`, we execute the command below in the terminal.

```Python
import sounddevice as sd

> ```

Reviewer: Are you sure about this? You can't execute an import command in the terminal :) Maybe you mean installing it using pip in the terminal?

We use methods such as Universal threshold, Bayes, and Sure minimax in setting the threshold.

### Scikit-image python package
This package provides flexible routines of image processing. It is written in the python language. In addition, it has a function library for wavelet-based denoising under restoration. 

Although it is mainly applicable for 2-D images, it can use it for 1-D signals. The good news is that the `scikit` image package is already available in anaconda; thus no need of installing it.

`Scikit` wavelet denoising includes two main functions. These functions are `estimate_sigma()` and `denoise_wavelet()`. 

For thresholding estimation, there are only two supported methods.

- Universal threshold(VishuShrink).
- Bayes Shrink(this is the default method).

By default  the `Denoise_wavelet()` function is used  for images. However, in this tutorial, we will use it for 1-D signals, meaning that the attributes related to the image are dropped. The syntax for using this function is:

```python
y = denoise_wavelet(x, wavelet='db1', mode='soft', wavelet_levels=n, method='BayesShrink', rescale_sigma='True')

> ```

Author: This is just a denoising syntax. You use it in your program just as we have done.

```Python
x = pywt.data.ecg().astype(float)/256  # In-built ecg data is read
```

The signal obtained from the database is noise-free. We need to add noise to it to perform the denoising operation. To add the noise, we first define the noise variance. The noise variance is the noise energy per sample.

```python
sigma = 0.05  # Noise variance
x_noisy = x + sigma * np.random.randn(x.size)   # Adding noise to signal
```

To add the noise to our signal(x), we use the formula `x + sigma * np.random.randn(x.size)`. 

The `randn()` is used to add a random noise in consideration to the noise variance `sigma`.

Let us now perform wavelet denoising using the `denoise_wavelet()` function using the syntax described earlier, as shown below:

```python
# Wavelet denoising
x_denoise = denoise_wavelet(x_noisy, method='BayesShrink', mode='soft', wavelet_levels=3, wavelet='sym8', rescale_sigma='True')
```

We plot the noisy signal `x_noisy` and the denoised signal `x-denoise` for visualisation.

```python
plt.figure(figsize=(20, 10), dpi=100)
plt.plot(x_noisy)
plt.plot(x_denoise)

> ```

Reviewer 2: This code snippet won't show the image. You are missing the `plt.show()` snippet.

When plotting, we first create a figure using the `plt.figure()` function. This figure is of the size 20x10 pixel as described by the function property `figsize`. Next, the noisy and the denoised signals are plotted in the exact figure using the `plt.plot()` function. The output of this plot is as shown below:

![Plot of the signals](/engineering-education/wavelet-transform-analysis-of-1d-signals-using-python/wavelet-two.png)

The blue signal is the noisy ecg signal, and the red is the denoised signal. Let us now look at how to denoise the audio signal since it is also a sample of a 1-D signal.

### Python code for denoising an audio signal
The first step is importing the libraries that we are going to use.

```Python
from scipy.io import wavfile
import numpy as np
from skimage.restoration import denoise_wavelet
import matplotlib.pyplot as plt

> ```

Reviewer 2: I would suggest you do all your imports once.

Author: These are two different programs. So they are done separately and that applies to the package installations as well.

We are importing the `wavfile` from the `scipy.io`. It is used to read audio signals in the `.wav` format. Since some numerical operations are involved, we import `numpy`. We also import the `denoise_wavelet()` function and `matplotlib.pyplot`.

> We then read the audio file using `wavfile.read()` function and normalize the amplitude as shown below:

Reviewer 2: We have already discussed these above.

```python
Fs, x = wavfile.read('guitar.wav') # Reading audio wave file
x = x/max(x)   # Normalizing amplitude

> ```

Reviewer 2: ```bash
fid = open(filename, 'rb')
FileNotFoundError: [Errno 2] No such file or directory: 'guitar.wav'
```

> We are reading the `guitar.wav` in the current directory. The signal samples will be stored in the `x` variable and the sampling frequency to `Fs` variable. 

Reviewer 2: You did not mention where to find this file and how to add it to the project.

The amplitude is normalized because `wavfile` reads the audio in `int16` mode. This mode gives large values. Therefore, we need to reduce these values through normalization.

Since the audio signal has no noise, we add noise similar to what we did for the ecg signal.

```python
sigma = 0.05  # Noise variance
x_noisy = x + sigma * np.random.randn(x.size)   # Adding noise to signal

# Wavelet denoising
x_denoise = denoise_wavelet(x_noisy, method='VisuShrink', mode='soft', wavelet_levels=3, wavelet='sym8', rescale_sigma='True')
```

After adding the random noise, we denoise the signal using the `denoise_wavelet()` function. Here, we use the `VishuShrink` method. 
For visualization, we plot the output.

```python
plt.figure(figsize=(20, 10), dpi=100)
plt.plot(x_noisy)
plt.plot(x_denoise)
plt.show()
```

![Output for the denoised audio signal](/engineering-education/wavelet-transform-analysis-of-1d-signals-using-python/wavelet-three.png)

> The blue signal is the noisy signal, and the orange is the denoised output. Now, it is hard to know whether the denoising is done or not. It can be done by either calculating the peak signal to noise ratio(PSNR) or physically listening to the music. 

Reviewer 2: ```suggestion
The blue signal is the noisy signal, while the orange is the denoised output. It is hard to know whether the denoising is done or not. 

However, this can be done by either calculating the peak signal to noise ratio(PSNR) or physically listening to the music. 
```
